File: lib/src/main/java/com/unity3d/ads/webview/bridge/WebViewBridgeInterface.java
package com.unity3d.ads.webview.bridge;

import android.webkit.JavascriptInterface;

import com.unity3d.ads.log.DeviceLog;

import org.json.JSONArray;
import org.json.JSONException;

public class WebViewBridgeInterface {
	@JavascriptInterface
	public void handleInvocation(String data) throws JSONException {
		DeviceLog.debug("handleInvocation " + data);

		JSONArray invocationArray = new JSONArray(data);
		Invocation batch = new Invocation();

		for (int idx = 0; idx < invocationArray.length(); idx++) {
			JSONArray currentInvocation = (JSONArray)invocationArray.get(idx);
			String className = (String)currentInvocation.get(0);
			String methodName = (String)currentInvocation.get(1);
			JSONArray parameters = (JSONArray)currentInvocation.get(2);
			String callback = (String)currentInvocation.get(3);

			batch.addInvocation(className, methodName, getParameters(parameters), new WebViewCallback(callback, batch.getId()));
		}

		for (int idx = 0; idx < invocationArray.length(); idx++) {
			batch.nextInvocation();
		}

		batch.sendInvocationCallback();
	}

	@JavascriptInterface
	public void handleCallback(String callbackId, String callbackStatus, String rawParameters) throws Exception {
		DeviceLog.debug("handleCallback " + callbackId + " " + callbackStatus + " " + rawParameters);

		JSONArray parametersJsonArray = new JSONArray(rawParameters);
		Object[] parameters = null;
		if(parametersJsonArray.length() > 0) {
			parameters = new Object[parametersJsonArray.length()];
			for(int i = 0; i < parametersJsonArray.length(); ++i) {
				parameters[i] = parametersJsonArray.get(i);
			}
		}

		WebViewBridge.handleCallback(callbackId, callbackStatus, parameters);
	}

	private Object[] getParameters (JSONArray parametersJson) throws JSONException {
		Object[] parameters = new Object[parametersJson.length()];
		for(int i = 0; i < parametersJson.length(); ++i) {
			parameters[i] = parametersJson.get(i);
		}

		return parameters;
	}
}

File: unity-ads/src/main/java/com/unity3d/services/core/webview/bridge/WebViewBridgeInterface.kt
package com.unity3d.services.core.webview.bridge

import android.net.Uri
import android.webkit.JavascriptInterface
import android.webkit.WebView
import androidx.webkit.JavaScriptReplyProxy
import androidx.webkit.WebMessageCompat
import com.unity3d.services.core.log.DeviceLog
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.asExecutor
import org.json.JSONArray
import org.json.JSONObject

class WebViewBridgeInterface(
    private val webViewBridge: IWebViewBridge = SharedInstances.webViewBridge,
    private val webViewAppInvocationCallbackInvoker: IInvocationCallbackInvoker = SharedInstances.webViewAppInvocationCallbackInvoker,
) {
    private fun JSONArray.toTypedArray() = (0 until length()).map(::get).toTypedArray()

    @JavascriptInterface
    fun handleInvocation(data: String) {
        DeviceLog.debug("handleInvocation $data")
        val invocationArray = JSONArray(data)
        val batch = Invocation(webViewAppInvocationCallbackInvoker, webViewBridge)

        for (idx in 0 until invocationArray.length()) {
            val currentInvocation = invocationArray[idx] as JSONArray
            val className = currentInvocation[0] as String
            val methodName = currentInvocation[1] as String
            val parameters = currentInvocation[2] as JSONArray
            val callback = currentInvocation[3] as String

            batch.addInvocation(className, methodName, parameters.toTypedArray(), WebViewCallback(callback, batch.id))
            batch.nextInvocation()
        }

        batch.sendInvocationCallback()
    }

    @JavascriptInterface
    fun handleCallback(callbackId: String, callbackStatus: String, rawParameters: String) {
        DeviceLog.debug("handleCallback $callbackId $callbackStatus $rawParameters")
        val parameters = JSONArray(rawParameters)

        webViewBridge.handleCallback(callbackId, callbackStatus, parameters.toTypedArray())
    }

    fun onHandleInvocation(view: WebView, message: WebMessageCompat, sourceOrigin: Uri, isMainFrame: Boolean, replyProxy: JavaScriptReplyProxy) {
        val data = message.data

        if (!isMainFrame || data.isNullOrBlank()) return

        handleInvocation(data)
    }

    fun onHandleCallback(view: WebView, message: WebMessageCompat, sourceOrigin: Uri, isMainFrame: Boolean, replyProxy: JavaScriptReplyProxy) {
        val data = message.data

        if (!isMainFrame || data.isNullOrBlank()) return

        val parameters = JSONObject(data)

        val callbackId = parameters.getString("id")
        val callbackStatus = parameters.getString("status")
        val rawParameters = parameters.getString("parameters")

        handleCallback(callbackId, callbackStatus, rawParameters)
    }
}
